index.d.ts

```diff
@@ -0,0 +1,138 @@
+    (req: RoutedRequest, res: any, next: NextFunction): any;
+  }
+
+  interface RequestHandler {
+    // tslint:disable-next-line callable-types (This is extended from and can't extend from a type alias in ts<2.2
+    (req: IncommingRequest, res: any, next: NextFunction): any;
+  }
+
+
+  interface NextFunction {
+    // tslint:disable-next-line callable-types (In ts2.1 it thinks the type alias has no call signatures)
+    (err?: any): void;
+  }
+
+  type ErrorRequestHandler = (err: any, req: IncommingRequest, res: any, next: NextFunction) => any;
+
+  type PathParams = string | RegExp | Array<string | RegExp>;
+
+  type RequestHandlerParams = RouteHandler | ErrorRequestHandler | Array<RouteHandler | ErrorRequestHandler>;
+
+  interface IRouterMatcher<T> {
+    (path: PathParams, ...handlers: RouteHandler[]): T;
+    (path: PathParams, ...handlers: RequestHandlerParams[]): T;
+  }
+
+  interface IRouterHandler<T> {
+    (...handlers: RouteHandler[]): T;
+    (...handlers: RequestHandlerParams[]): T;
+  }
+
+  interface IRouter {
+    /**
+     * Map the given param placeholder `name`(s) to the given callback(s).
+     *
+     * Parameter mapping is used to provide pre-conditions to routes
+     * which use normalized placeholders. For example a _:user_id_ parameter
+     * could automatically load a user's information from the database without
+     * any additional code,
+     *
+     * The callback uses the samesignature as middleware, the only differencing
+     * being that the value of the placeholder is passed, in this case the _id_
+     * of the user. Once the `next()` function is invoked, just like middleware
+     * it will continue on to execute the route, or subsequent parameter functions.
+     *
+     *      app.param('user_id', function(req, res, next, id){
+     *        User.find(id, function(err, user){
+     *          if (err) {
+     *            next(err);
+     *          } else if (user) {
+     *            req.user = user;
+     *            next();
+     *          } else {
+     *            next(new Error('failed to load user'));
+     *          }
+     *        });
+     *      });
+     */
+    param(name: string, handler: RequestParamHandler): this;
+
+    /**
+     * Alternatively, you can pass only a callback, in which case you have the opportunity to alter the app.param()
+     *
+     * @deprecated since version 4.11
+     */
+    param(callback: (name: string, matcher: RegExp) => RequestParamHandler): this;
+
+    /**
+     * Special-cased "all" method, applying the given route `path`,
+     * middleware, and callback to _every_ HTTP method.
+     */
+    all: IRouterMatcher<this>;
+
+    get: IRouterMatcher<this>;
+    post: IRouterMatcher<this>;
+    put: IRouterMatcher<this>;
+    delete: IRouterMatcher<this>;
+    patch: IRouterMatcher<this>;
+    options: IRouterMatcher<this>;
+    head: IRouterMatcher<this>;
```

Comment:
There seems to be a lot more methods missing here. Can you add the rest? Examples are `.bind`, `.propfind`, etc.

Reply:
`.bind` and all the default object level props you get for free. I can't find `.propfind` in the codebase. Not on `2.0` either?

Reply:
All these methods are not in the code base -- they are dynamic depending on the Node.js version -- even the ones like `.options` is not in the code base at all...

We iterate over https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_http_methods and add the lower case as a method to the router prototype. Here is propfind existing:

```
$ node -pe 'require("router")().propfind'
[Function]
```